tracker/enqueue.go

```diff
@@ -158,12 +158,13 @@ func (i *impl) TrackReference(ref Reference, obj interface{}) error {
 			// The simplest way of eliminating such a window is to call the
 			// callback to "catch up" immediately following new
 			// registrations.
-			i.cb(key)
+			defer i.cb(key)
```

Comment:
Would it be better to have a "locked" get func that reruns nil or cb. Then the code here can be simplified and the unlocks won't be spread around

Reply:
I don't know what you mean?

Reply:
well, first of all, @vagababov do you agree that we should not be holding the lock while making callbacks. If I'm just making up dumb things, then this change is moot.

Reply:
As a general approach, I do agree with that, since god knows what the callback might do.
What I meant here is that the unlock and lock are now far from each other. I'd prefer we have a "locked" function that returns the callback and then the code is
```
lock
defer getCB()
unlock
```

Reply:
🤔 I would expect that only the invocation (and not the read of `i.cb`) happens outside of the lock, but that is perhaps incorrect...

Reply:
Are you suggesting that this is why `race` is freaking out?

Reply:
I haven't looked at the race errors, let me check.
But in general the further the unlock is from lock, the bigger the chance some path will forget to unlock it, or some return/panic will escape unlocking it.

Fun story: on one of my GCP projects we were seeing runaway latencies on some instances.
Turned out that we were using LRU map whose `Get` was not readlock safe (which is kind of reasonable, given you have to update the last used timestamp).
And we were using `lock() x unlock()` pattern, where x would include map read. Which would sometimes panic and keep the mutex locked and the whole system would then grind to a halt. 

Here you have 71 lines of separation between lock and unlock 133-204 (well each path is shorter, but I'll exaggerate) or 227-274, each of the code lines in theory can panic.
So I'd make the code along the lines of:
```
func acquireCB() func(namespacedname){
  lock()
  defer unlock()
  if selector == nil {...return x}
  ...
 return y
}

func (i *impl) OnChanged(obj interface{}) {
  ...
  cb := acquireCB(...)
  cb()
}
```
Same for track reference.

Reply:
I've rewritten this as something much simpler.